worker/gouging.go

```diff
@@ -87,113 +123,238 @@ func (gr GougingResults) CanForm() []error {
+	}
+
+	// check LatestRevisionCost - expect sane value
+	bw, overflow := pt.DownloadBandwidthCost.Mul64WithOverflow(modules.SectorSize)
 	if overflow {
-		return fmt.Errorf("overflow detected when computing upload price per TiB")
+		return errors.New("DownloadBandwidthCost overflows LatestRevisionCost calculation")
 	}
-	uploadPriceTotalShards, overflow := upptb.Mul64WithOverflow(uint64(totalShards))
+	cost, overflow := bw.AddWithOverflow(pt.InitBaseCost)
 	if overflow {
-		return fmt.Errorf("overflow detected when multiplying %v * %v in upload gouging", upptb, totalShards)
+		return errors.New("InitBaseCost overflows LatestRevisionCost calculation")
 	}
-	uploadPrice := uploadPriceTotalShards.Div64(uint64(minShards))
-	if !gs.MaxUploadPrice.IsZero() && uploadPrice.Cmp(gs.MaxUploadPrice) > 0 {
-		return fmt.Errorf("cost per TiB exceeds max ul price: %v>%v", uploadPrice, gs.MaxUploadPrice)
+	if pt.LatestRevisionCost.Cmp(cost) > 0 {
+		return fmt.Errorf("LatestRevisionCost of %v exceeds maximum cost of %v", pt.SubscriptionNotificationCost, cost)
 	}
 
-	return nil
-}
-
-func checkFormContractGouging(gs api.GougingSettings, hs rhpv2.HostSettings) error {
-	// check base rpc price
-	if !gs.MaxRPCPrice.IsZero() && hs.BaseRPCPrice.Cmp(gs.MaxRPCPrice) > 0 {
-		return fmt.Errorf("rpc price exceeds max: %v>%v", hs.BaseRPCPrice, gs.MaxRPCPrice)
+	// check RenewContractCost - expect 100nS default
+	if types.Siacoins(1).Mul64(100).Div64(1e9).Cmp(pt.RenewContractCost) < 0 {
+		return fmt.Errorf("RenewContractCost of %v exceeds 100nS", pt.RenewContractCost)
 	}
 
-	// check contract price
-	if !gs.MaxContractPrice.IsZero() && hs.ContractPrice.Cmp(gs.MaxContractPrice) > 0 {
-		return fmt.Errorf("contract price exceeds max: %v>%v", hs.ContractPrice, gs.MaxContractPrice)
+	// check RevisionBaseCost - expect 0H default
+	if types.ZeroCurrency.Cmp(pt.RevisionBaseCost) < 0 {
+		return fmt.Errorf("RevisionBaseCost of %v exceeds 0H", pt.RevisionBaseCost)
 	}
 
-	// check MaxCollateral
-	if hs.MaxCollateral.IsZero() {
-		return errors.New("MaxCollateral of host is 0")
+	// check block height
+	if !cs.Synced {
+		if pt.HostBlockHeight < cs.BlockHeight {
+			return fmt.Errorf("consensus not synced and host block height is lower, %v < %v", pt.HostBlockHeight, cs.BlockHeight)
+		}
+	} else {
+		if !(cs.BlockHeight-blockHeightLeeway <= pt.HostBlockHeight && pt.HostBlockHeight <= cs.BlockHeight+blockHeightLeeway) {
+			return fmt.Errorf("consensus is synced and host block height is not within range, %v %v %v", cs.BlockHeight, pt.HostBlockHeight, blockHeightLeeway)
+		}
 	}
-	if hs.MaxCollateral.Cmp(gs.MinMaxCollateral) < 0 {
-		return fmt.Errorf("MaxCollateral is below minimum: %v<%v", hs.MaxCollateral, gs.MinMaxCollateral)
+
+	// check TxnFeeMaxRecommended - expect at most a multiple of our fee
+	if !txnFee.IsZero() && pt.TxnFeeMaxRecommended.Cmp(txnFee.Mul64(5)) > 0 {
+		return fmt.Errorf("TxnFeeMaxRecommended %v exceeds %v", pt.TxnFeeMaxRecommended, txnFee.Mul64(5))
 	}
```

Comment:
I think you missed `TxnFeeMinRecommended`.

Reply:
```go
func (tp txpool) RecommendedFee() (fee types.Currency) {
	_, max := tp.tp.FeeEstimation()
	convertToCore(&max, &fee)
	return
}
```

Reply:
I added an assertion to ensure it's lower than the max fee